File: 61_search-for-a-range/search-for-a-range.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/search-for-a-range
@Language: Python
@Datetime: 16-05-17 22:06
'''

class Solution:
    """
    @param A : a list of integers
    @param target : an integer to be searched
    @return : a list of length 2, [index1, index2]
    """
    def searchRange(self, A, target):
        # write your code here
        # corner case
        if A is None or len(A) == 0:
            return [-1, -1]
        start = 0
        end = len(A) - 1
        while start + 1 < end:
            mid = start + (end - start) / 2
            if A[mid] < target:
                start = mid + 1
            else:
                end = mid
        if A[start] == target:
            lowerbound = start
        elif A[end] == target:
            lowerbound = end
        else:
            return [-1, -1]
        
        start = lowerbound
        end = len(A) - 1
        while start + 1 < end:
            mid = start + (end - start) / 2
            if A[mid] > target:
                end = mid - 1
            else:
                start = mid
        if A[end] == target:
            upperbound = end
        else:
            upperbound = start
        
        return [lowerbound, upperbound]
        
            
            
        
        

File: 61_search-for-a-range/search-for-a-range.java
/*
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/search-for-a-range
@Language: Java
@Datetime: 16-05-18 03:09
*/

public class Solution {
    /** 
     *@param A : an integer sorted array
     *@param target :  an integer to be inserted
     *return : a list of length 2, [index1, index2]
     */
    public int[] searchRange(int[] A, int target) {
        //write your code here
        if (A == null || A.length == 0) {
            return new int[]{-1, -1}; 
        }
        
        int start = 0;
        int end = A.length - 1;
        int[] range = {-1, -1};
        
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (A[mid] < target) {
                start = mid;
            }
            else {
                end = mid;
            }
        }
        if (A[start] == target) {
            range[0] = start;
        }
        else if (A[end] == target) {
            range[0] = end;
        }
        else {
            return range;
        }
        
        start = range[0];
        end = A.length - 1;
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (A[mid] > target) {
                end = mid;
            }
            else {
                start = mid;
            }
        }
        if (A[end] == target) {
            range[1] = end;
        }
        else {
            range[1] = start;
        }
        return range;
    }
}
